Math/NumberTheory/Primes/Factorisation/Montgomery.hs

```diff
@@ -152,75 +155,88 @@ stdGenFactorisation primeBound sg digits n
-                                         else repFact cof b1 b2 (count-1)
-                            return (merge dp cp, dc ++ cc)
-                        g -> do
-                            let d' = d `quot` g
-                                c' = cof `quot` g
-                            (dp,dc) <- if ptest d'
-                                         then return ([(d',1)],[])
-                                         else repFact d' b1 b2 (count-1)
-                            (cp,cc) <- if ptest c'
-                                         then return ([(c',1)],[])
-                                         else repFact c' b1 b2 (count-1)
-                            (gp,gc) <- if ptest g
-                                         then return ([(g,2)],[])
-                                         else dbl <$> repFact g b1 b2 (count-1)
-                            return  (mergeAll [dp,cp,gp], dc ++ cc ++ gc)
+
+        ptest :: Integer -> Bool
+        ptest = maybe primeTest (\bd k -> k <= bd || primeTest k) primeBound
+
+        rndR :: Integer -> State g Integer
+        rndR k = state (prng k)
+
+        perfPw :: Integer -> (Integer, Int)
+        perfPw = maybe highestPower (largePFPower . integerSquareRoot') primeBound
+
+        fact :: Integer -> Int -> State g [(Integer, Int)]
+        fact 1 _ = return mempty
+        fact m digs = do
+          let (b1, b2, ct) = findParms digs
+          Factors pfs cfs <- repFact m b1 b2 ct
+          case cfs of
+            [] -> return pfs
+            _  -> do
+              nfs <- forM cfs $ \(k, j) ->
+                  map (second (* j)) <$> fact k (if null pfs then digs + 5 else digs)
+              return $ sconcat $ pfs :| nfs
+
+        repFact :: Integer -> Word -> Word -> Word -> State g Factors
+        repFact 1 _ _ _ = return mempty
+        repFact m b1 b2 count =
+          case perfPw m of
+            (_, 1) -> workFact m b1 b2 count
+            (b, e)
+              | ptest b   -> return $ Factors [(b, e)] []
+              | otherwise -> modifyPowers (* e) <$> workFact b b1 b2 count
+
+        workFact :: Integer -> Word -> Word -> Word -> State g Factors
+        workFact 1 _ _ _ = return mempty
+        workFact m _ _ 0 = return $ Factors [] [(m, 1)]
+        workFact m b1 b2 count = do
+          s <- rndR m
+          case s `modulo` fromInteger m of
+            InfMod{} -> error "impossible case"
+            SomeMod sm -> case montgomeryFactorisation b1 b2 sm of
+              Nothing -> workFact m b1 b2 (count - 1)
+              Just d  -> do
+                let cs = splitIntoCoprimes [(d, 1), (m `quot` d, 1)]
+                fmap mconcat $ flip mapM cs $
+                  \(x, xm) -> if ptest x
+                              then pure (Factors [(x, xm)] [])
+                              else repFact x b1 b2 (count - 1)
+
+data Factors = Factors
+  { _primeFactors     :: [(Integer, Int)]
+  , _compositeFactors :: [(Integer, Int)]
+  }
+
```

Comment:
the safety/correctness condition for this semigroup seems tricky to enforce outside of internals, eg if 2 is a prime factor and 4 is a composite factor, right?

Reply:
(granted any checking / soundness guarantees would be bad for algorithm complexity)

Reply:
In fact, this `Semigroup` is lawful in any case, it just may be less suitable for factorisation. I have moved comments to the relevant pieces of `curveFactorisation`.